### PR TITLE
Cleanup terraformer library

### PR DESCRIPTION
**Pull request by timebertt**
**How to categorize this PR?**

/area logging quality
/kind cleanup
/priority normal

**What this PR does / why we need it**:

This PR cleans up several things related to the terraformer library:
- removes deprecated functions: `Terraformer.SetVariablesEnvironment` and `GenerateVariablesEnvironment` (⚠️ breaking change)
- cleans up `context.TODO`s and replace them by proper contexts passed by callers (⚠️ breaking change)
- improve the logs in the infrastructure controller/reconciler similar to https://github.com/gardener/gardener/pull/2807
- switch to `logr` in the terraformer, as all the extensions use it everywhere and using `logrus` only there makes the logs inconsistent and hard to read (⚠️ breaking change)

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```breaking dependency
The deprecated functions in the terraformer library (`SetVariablesEnvironment` and `GenerateVariablesEnvironment`) have been removed.
```
```breaking dependency
The `Terraformer` functions have been changed to allow passing proper contexts. Please adapt your usage accordingly.
```
```breaking dependency
The terraformer library was switched to `logr` instead of `logrus` in order to have more consistent and readable logging in the infrastructure controllers of provider extensions. Please adapt your usage accordingly.
```

### CODE REVIEW

File: extensions/pkg/terraformer/terraformer.go [hunk 1]
```diff
@@ -178,50 +183,56 @@ func (t *terraformer) execute(ctx context.Context, command string) error {
...
 
 	if !skipApplyOrDestroyPod {
 		// Create Terraform Pod which executes the provided command
 		generateName := t.computePodGenerateName(command)
+
+		logger.Info("Deploying Terraformer Pod", "generateName", generateName)
 		pod, err := t.deployTerraformerPod(ctx, generateName, command)
 		if err != nil {
-			return fmt.Errorf("failed to deploy the Terraformer Pod with .meta.generateName '%s': %s", generateName, err.Error())
+			return fmt.Errorf("failed to deploy the Terraformer Pod with .meta.generateName %q: %w", generateName, err)
```
Comment (vpnachev) [inline extensions/pkg/terraformer/terraformer.go:196]: Why are you switching to `%w`? It is not looking very good to me.

```golang
package main

import (
	"fmt"
)

func main() {
	const name, id = "bueller", 17
	err := fmt.Errorf("user %q (id %d) not found", name, id)
	fmt.Printf("%w\n", err) // %!w(*errors.errorString=&{user "bueller" (id 17) not found})
	fmt.Printf("%v\n", err) // user "bueller" (id 17) not found
}
```
The above code snippet is taken from [here](https://golang.org/pkg/fmt/#Errorf)

File: extensions/pkg/terraformer/waiter.go [hunk 2]
```diff
@@ -19,10 +19,12 @@ import (
 	"fmt"
 	"time"
 
-	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
-	"github.com/gardener/gardener/pkg/utils/retry"
+	"github.com/go-logr/logr"
 	corev1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
```
Comment (vpnachev) [inline extensions/pkg/terraformer/waiter.go:25]:
```suggestion
```

**Review by vpnachev — COMMENTED** [review]
Looks very good to me. 
I have just one comment about formatting.

File: extensions/pkg/terraformer/terraformer.go (hunk 1 above)
Comment (timebertt) [reply extensions/pkg/terraformer/terraformer.go:196]: `%w` is only working with `fmt.Errorf`: https://golang.org/pkg/fmt/#Errorf
That's why it's looking strange with `fmt.Printf` (behaves like `%+v`, I believe).

E.g.:
```go
const name, id = "bueller", 17
err := fmt.Errorf("user %q (id %d) not found", name, id)
fmt.Println(fmt.Errorf("wrapped error: %w", err)) // wrapped error: user "bueller" (id 17) not found
```

Comment (kon-angelo) [reply extensions/pkg/terraformer/terraformer.go:196]: Well the printing should be the same. Snippet for reference:

```go
package main

import (
	"fmt"
)

func main() {
	const name, id = "bueller", 17
	err := fmt.Errorf("user %q (id %d) not found", name, id)
	errOut1:= fmt.Errorf("%v",err)
	errOut2:= fmt.Errorf("%w",err)
	fmt.Printf("%v\n", errOut1) //prints user "bueller" (id 17) not found
	fmt.Printf("%v\n", errOut2) //user "bueller" (id 17) not found
}
```

@timebertt You should use `%w` only if you  want the callers to have access to the original error. If you think that the callers cannot use the internal error then wrap it with `%v`

```go
return fmt.Errorf("failed to deploy the Terraformer Pod with .meta.generateName %q: %v", generateName, err)
```

Comment (vpnachev) [reply extensions/pkg/terraformer/terraformer.go:196]: Thanks for the explanations :heart: .

Comment (timebertt) [reply extensions/pkg/terraformer/terraformer.go:196]: > @timebertt You should use `%w` only if you want the callers to have access to the original error. If you think that the callers cannot use the internal error then wrap it with `%v`

I guess, in general it makes sense to wrap errors, that will be propagated up to other components (as it's the case here).
They might decide to actually unwrap the errors at some point.

**Review by vpnachev — APPROVED** [review]
/lgtm
/needs second-opinion

**Review by kon-angelo — APPROVED** [review]

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**